### ISSUE TITLE
[BUG] initialization issue for struct orc_read_arg, pq_read_arg for Power Platform/gcc 7.3

### ISSUE THREAD

**Issue by pradghos** [rapidsai/cudf#1878]
Problem : 

**Describe the bug**
File location : `python/cudf/bindings/parquet.pxd` and `python/cudf/bindings/orc.pxd`
`struct orc_read_arg` and `struct pq_read_arg` in cython are not getting initialized properly while `skip_rows` / `use columns` are passed as `None`. and multiple test cases are failing from `test_parquet.py` and `test_orc.py`

Test example:

**test case ** 
test_parquet.py

test_parquet_reader_basic[None-10-columns0-cudf]
`RuntimeError: cuDF failure at: /opt/anaconda3/conda-bld/libcudf_1559038627449/work/cpp/src/io/parquet/parquet_reader.cu:205: Invalid row start` 

in the code - `CUDF_EXPECTS(row_start <= get_total_rows(), "Invalid row start");`  

**test case ** 
test_orc.py
```
test_orc.py::test_orc_reader_basic[orc_args0-cudf] FAILED                                                                                                     [ 13%]
test_orc.py::test_orc_reader_basic[orc_args1-cudf] FAILED                                                                                                     [ 26%]
test_orc.py::test_orc_reader_decimal FAILED                                                                                                                   [ 33%]

test_orc.py::test_orc_read_rows[0-1] Segmentation fault (core dumped)
```
issue :1 
```
(gdb) where
#0  0x00007fffb4fd713c in __strlen_power8 () from /usr/lib64/libc.so.6
#1  0x00007ffda061fdd8 in read_orc () from /opt/anaconda3/lib/python3.6/site-packages/cudf/bindings/../../../../libcudf.so
#2  0x00007ffd943fbc30 in __pyx_f_4cudf_8bindings_3orc_cpp_read_orc(_object*, int, __pyx_opt_args_4cudf_8bindings_3orc_cpp_read_orc*) [clone .isra.47] [clone .constprop.51] () from /opt/anaconda3/lib/python3.6/site-packages/cudf/bindings/orc.cpython-36m-powerpc64le-linux-gnu.so
#3  0x00007ffd94401650 in __pyx_pw_4cudf_8bindings_3orc_3cpp_read_orc(_object*, _object*, _object*) ()
   from /opt/anaconda3/lib/python3.6/site-packages/cudf/bindings/orc.cpython-36m-powerpc64le-linux-gnu.so
#4  0x000000013ca84f6c in _PyCFunction_FastCallDict (func_obj=0x7ffd967b50d8, args=<optimized out>, nargs=4, kwargs=0x0)
    at /root/worker/containers/000003ef0p3/tmp/build/80754af9/python_1546130315243/work/Objects/methodobject.c:231
#5  0x000000013cad1334 in _PyCFunction_FastCallKeywords (func=<optimized out>, stack=<optimized out>, nargs=<optimized out>, kwnames=<optimized out>)
    at /root/worker/containers/000003ef0p3/tmp/build/80754af9/python_1546130315243/work/Objects/methodobject.c:294
#6  0x000000013cb6e32c in call_function (pp_stack=0x7fffd87933f8, oparg=4, kwnames=0x0)

```
issue:2 
```
E   RuntimeError: cuDF failure at: /opt/anaconda3/conda-bld/libcudf_1559038627449/work/cpp/src/io/orc/orc_reader.cu:168: Invalid row start
                 
```
in the code - `CUDF_EXPECTS(row_start <= get_total_rows(), "Invalid row start");`

I have tried printing the value in cython structure when skip_rows and column is None.  

```
orc_reader.skip_rows:32765,  === > which is probably garbage value
orc_reader.num_rows -1
usecols is None
orc_reader.use_cols_len 1882049728  == > garbage value
```
**Steps/Code to reproduce bug**

running the cuDF pytests on Power Platform. 

**Expected behavior**
All the cython structure member should correctly initialized. 

**Environment details (please complete the following information):**
 `cuDF 0.7.2`
 `Power Platform`
 `gcc 7.3`

**Comment by pradghos** [comment]
Below Fixes are to address the issue -

```
diff --git a/python/cudf/bindings/orc.pyx b/python/cudf/bindings/orc.pyx
index a26a016..aaaed4a 100644
--- a/python/cudf/bindings/orc.pyx
+++ b/python/cudf/bindings/orc.pyx
@@ -56,17 +56,23 @@ cpdef cpp_read_orc(path, columns=None, skip_rows=None, num_rows=None):
     orc_reader.source = source_ptr
 
     usecols = columns
+    arr_cols = []
+    orc_reader.use_cols_len = 0
+
     cdef vector[const char*] vector_use_cols
     if usecols is not None:
-        arr_cols = []
         for col in usecols:
             arr_cols.append(str(col).encode())
-        vector_use_cols = arr_cols
-        orc_reader.use_cols = vector_use_cols.data()
         orc_reader.use_cols_len = len(usecols)
 
+    vector_use_cols = arr_cols
+    orc_reader.use_cols = vector_use_cols.data()
+
     if skip_rows is not None:
         orc_reader.skip_rows = skip_rows
+    else:
+        orc_reader.skip_rows = 0
+
     if num_rows is not None:
         orc_reader.num_rows = num_rows
     else:
diff --git a/python/cudf/bindings/parquet.pyx b/python/cudf/bindings/parquet.pyx
index 205586f..bf29e54 100644
--- a/python/cudf/bindings/parquet.pyx
+++ b/python/cudf/bindings/parquet.pyx
@@ -50,15 +50,18 @@ cpdef cpp_read_parquet(path, columns=None, row_group=None, skip_rows=None,
     pq_reader.source = source_ptr
 
     usecols = columns
+    arr_cols = []
+    pq_reader.use_cols_len = 0
+
     cdef vector[const char*] vector_use_cols
     if usecols is not None:
-        arr_cols = []
         for col in usecols:
             arr_cols.append(str(col).encode())
-        vector_use_cols = arr_cols
-        pq_reader.use_cols = vector_use_cols.data()
         pq_reader.use_cols_len = len(usecols)
 
+    vector_use_cols = arr_cols
+    pq_reader.use_cols = vector_use_cols.data()
+
     if row_group is not None:
         pq_reader.row_group = row_group
     else:
@@ -66,6 +69,9 @@ cpdef cpp_read_parquet(path, columns=None, row_group=None, skip_rows=None,
 
     if skip_rows is not None:
         pq_reader.skip_rows = skip_rows
+    else:
+        pq_reader.skip_rows = 0
+
     if num_rows is not None:
         pq_reader.num_rows = num_rows
else:
``` 
@jrhemstad @kkraus14 @harrism : Please let me know If I can create PR for the same. Thanks!

**Comment by kkraus14** [comment]
> Below Fixes are to address the issue -
> 
> ```
> diff --git a/python/cudf/bindings/orc.pyx b/python/cudf/bindings/orc.pyx
> index a26a016..aaaed4a 100644
> --- a/python/cudf/bindings/orc.pyx
> +++ b/python/cudf/bindings/orc.pyx
> @@ -56,17 +56,23 @@ cpdef cpp_read_orc(path, columns=None, skip_rows=None, num_rows=None):
>      orc_reader.source = source_ptr
>  
>      usecols = columns
> +    arr_cols = []
> +    orc_reader.use_cols_len = 0
> +
>      cdef vector[const char*] vector_use_cols
>      if usecols is not None:
> -        arr_cols = []
>          for col in usecols:
>              arr_cols.append(str(col).encode())
> -        vector_use_cols = arr_cols
> -        orc_reader.use_cols = vector_use_cols.data()
>          orc_reader.use_cols_len = len(usecols)
>  
> +    vector_use_cols = arr_cols
> +    orc_reader.use_cols = vector_use_cols.data()
> +
>      if skip_rows is not None:
>          orc_reader.skip_rows = skip_rows
> +    else:
> +        orc_reader.skip_rows = 0
> +
>      if num_rows is not None:
>          orc_reader.num_rows = num_rows
>      else:
> diff --git a/python/cudf/bindings/parquet.pyx b/python/cudf/bindings/parquet.pyx
> index 205586f..bf29e54 100644
> --- a/python/cudf/bindings/parquet.pyx
> +++ b/python/cudf/bindings/parquet.pyx
> @@ -50,15 +50,18 @@ cpdef cpp_read_parquet(path, columns=None, row_group=None, skip_rows=None,
>      pq_reader.source = source_ptr
>  
>      usecols = columns
> +    arr_cols = []
> +    pq_reader.use_cols_len = 0
> +
>      cdef vector[const char*] vector_use_cols
>      if usecols is not None:
> -        arr_cols = []
>          for col in usecols:
>              arr_cols.append(str(col).encode())
> -        vector_use_cols = arr_cols
> -        pq_reader.use_cols = vector_use_cols.data()
>          pq_reader.use_cols_len = len(usecols)
>  
> +    vector_use_cols = arr_cols
> +    pq_reader.use_cols = vector_use_cols.data()
> +
>      if row_group is not None:
>          pq_reader.row_group = row_group
>      else:
> @@ -66,6 +69,9 @@ cpdef cpp_read_parquet(path, columns=None, row_group=None, skip_rows=None,
>  
>      if skip_rows is not None:
>          pq_reader.skip_rows = skip_rows
> +    else:
> +        pq_reader.skip_rows = 0
> +
>      if num_rows is not None:
>          pq_reader.num_rows = num_rows
> else:
> ```
> 
> @jrhemstad @kkraus14 @harrism : Please let me know If I can create PR for the same. Thanks!

Please open the PR, it would be greatly appreciated, thanks!

**Comment by mjsamoht** [comment]
@j-ieong can you take a look at the proposed changes above?

If this is specific to Power platform then I guess we have no way to repro and confirm any fixes.

**Comment by pradghos** [comment]
@kkraus14 : Thanks for looking into the issue. I have created PR with the fixes.

**Comment by mjsamoht** [comment]
Fixed with #1887.
